### PR TITLE
NIFI-12707: Allow LookupRecord to operate on multiple child-records; …

### PR DESCRIPTION
**Pull request by markap14**
…in order to accommodate this, also needed to improve DataTypeUtils so that it knows that Record A is wider than Record B if Record A contains all fields of Record B and more.

# Summary

[NIFI-00000](https://issues.apache.org/jira/browse/NIFI-00000)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [ ] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [ ] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [ ] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [ ] Pull Request based on current revision of the `main` branch
- [ ] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [ ] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java [hunk 1]
```diff
@@ -1792,37 +1796,77 @@ public static Optional<DataType> getWiderType(final DataType thisDataType, final
...
+        for (final RecordField thisField : thisFields) {
+            if (!otherFields.contains(thisField)) {
+                allFieldsPresent = false;
+                break;
+            }
+        }
+
+        if (allFieldsPresent) {
+            return Optional.of(otherRecordDataType);
+        }
```
[the remark this answers is not in the document]
Comment (markap14) [reply nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java:1849]: Yes. Yes, it can :) Good catch.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for working on these improvements @markap14.

There are test failures in `TestLookupRecord` on the Windows build for the content-based comparison, due to newline differences in the pretty-printed JSON. There are some utility methods to normalize the differences.

**Comment by markap14** [comment]
Thanks @exceptionfactory have addressed.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/LookupRecord.java [hunk 2]
```diff
@@ -196,6 +198,18 @@ public class LookupRecord extends AbstractProcessor {
         .required(true)
         .build();
 
+    static final PropertyDescriptor RESULT_RECORD_PATH = new PropertyDescriptor.Builder()
+        .name("result-record-path")
+        .displayName("Result RecordPath")
```
Comment (dan-s1) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/LookupRecord.java:203]: Isn't the pattern for new properties to have the `.name` and `.displayName` be the same
```suggestion
        .name("Result RecordPath")
        .displayName("Result RecordPath")
```

Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/LookupRecord.java:203]: @dan-s1 Yes, but this is just moving the descriptor declaration `RESULT_RECORD_PATH`, so it isn't introducing a new property. See above changes on lines 158-163.

**Review by exceptionfactory — APPROVED** [review]
Thanks for the improvements @markap14, the test updates also look good. Thanks for the review as well @dan-s1. +1 merging

**Pull request closed without merging**